### ISSUE TITLE
Badges and NFTs

### ISSUE THREAD

**Issue by n13** [JoinSEEDS/seeds-smart-contracts#275]
There's different types of tokens we want to enable on Seeds

## Badges (DRAFT)
- Badges have a creator which is cryptographically secured (signer)
- Badges have a name and ID (unique ID)
- Badge has a max supply (can be unlimited)
- Badges can be created by anyone, but the interpretation of a badge is up to the interpreter
- Badges can be given to any account
- Badge can be transferable or non-transferable by the holder. Most will be non-transferable.
- An account can have 1 or more badges of the same kind? - Limit to 1, I think. 
- Removal of badges - to be defined
- Badge creator has 

## NFT
- transferable
- has creator
- limited supply, e.g. 1, 10, 20 - numbered. Or is it always 1.
- unique ID

**Comment by juliolrmonteiro** [comment]
I agree that scores should not be used as Badges.
You can have fungible things like scores or coins as **requisites** to earn or lose a badge, coded in smartcontracts.
I like to think of badges as markers or milestones, for human readable data analysis. So if you have X planted score, Y contribution score, Z community building score, you earn a AAA badge so everyone else can see you actually "beat that challenge". And you may lose the badge if the points drop below a threshold too.

**Comment by juliolrmonteiro** [comment]
They way I was thinking is that badges can be earned in 3 ways:

- Coded as a "trigger" on any smartcontract (like when an ORg receives X CS, they get the Reputable badge)
- Earned as reward from a Quest (you earn coins, and also "Reputation" or "Skill" badges as you complete the quest)
- Earned on any DHO (voted DHO badges) can be freely translated to SEEDS badges, by invoking an ACTION

This is cool because we can have "Nomination" public quests, that pay 0 SEEDS, but give someone a Badge. And we can also have private quests so that one individual can "recognize" another and give them a Badge.

**Comment by juliolrmonteiro** [comment]
Regarding losing badges, there's also 3 ways:

- Coded as a "trigger" on any smartcontract, if you reach below a threshold you can lose a badge
- Any Badge can be flagged by users, if the total cumulative flagging power (sum of REP) is greater than the REP power of the issuer(s), the user loses the badge.
- Badges can expire too, and need to be renewed

**Comment by juliolrmonteiro** [comment]
> An account can have 1 or more badges of the same kind? (NFTs)

Yes, this "kind" is usually called BadgeClass, so if HYPHA tokens were to become NFTs, each token share would then have the same BadgeClass, but still being unique. And you can count how many tokens of that same BadgeClass you own. This can also be seen as Badge "levels", for instance, if you have a "Coding skills" badge awarded to you once, it's level 1, but if you are awarded another badge with the same class, you can be seen as as having "Coding skills" level 2 now.

This is not very efficient for thousands or million of tokens, however, unless you use less tokens but each one has a different value (older ones more worthy, for example).

**Comment by juliolrmonteiro** [comment]
For reference, here's how [Mozilla OpenBadges](https://www.imsglobal.org/sites/default/files/Badges/OBv2p0Final/index.html) are defined:
```JSON
{
  "@context": "https://w3id.org/openbadges/v2",
  "id": "https://example.org/assertions/123",
  "type": "Assertion",
  "recipient": {
    "type": "email",
    "identity": "alice@example.org"
  },
  "issuedOn": "2016-12-31T23:59:59+00:00",
  "verification": {
    "type": "hosted"
  },
  "badge": {
    "type": "BadgeClass",
    "id": "https://example.org/badges/5",
    "name": "3-D Printmaster",
    "description": "This badge is awarded for passing the 3-D printing knowledge and safety test.",
    "image": "https://example.org/badges/5/image",
    "criteria": {
      "narrative": "Students are tested on knowledge and safety, both through a paper test and a supervised performance evaluation on live equipment"
    },
    "issuer": {
      "id": "https://example.org/issuer",
      "type": "Profile",
      "name": "Example Maker Society",
      "url": "https://example.org",
      "email": "contact@example.org",
      "verification": {
         "allowedOrigins": "example.org"
      }
    }
  }
}
```

**Comment by 7flash** [comment]
I support the idea to enable users create their own badges within ecosystem and distribute them. For example, some campaigns might give badges for planting trees, making donations, executing quests, etc...

**Comment by juliolrmonteiro** [comment]
After they are created only some fields can be changed.

Ways to lose badges:
- The issuer that created it can remove/expire it
- They can expire
- If the account is banned or deleted, what happens to the badges? (Nothing)

Any Resident account can create a badge (they become the **issuer**), just pay the fee.
Contract accounts don't pay the fee.

Have a **criteria** field (hashmap) to indicate how to earn/lose the badge. For example:, Citizen badge?
For quests, it can be the **id** of the Quest object that gave you that badge.

Badges scoped by user. Sorted by system badges first.

Badges as human readable milestones.

**Comment by n13** [comment]
## Badges Contract

### Tables
badge_definition (class)

- id
- issuer
- revoker
- time limit
- name (String)
- description (String)
- criteria (String)
- issuance_type (system / web of trust / individual / other)

OR

badge_link
id
account (index)
badge_id (index)


Create Badge
Issue Badge to ....

**Comment by n13** [comment]
Examples

issuer could be
dho.hypha
dho.seeds
or any region that has a DHO, etc
